ndcube/wcs/wrappers/resampled_wcs.py

```diff
@@ -13,34 +13,73 @@ class ResampledLowLevelWCS(BaseWCSWrapper):
             factor = [factor] * self.pixel_n_dim
-        self._factor = factor
+        self._factor = np.array(factor)
+        if len(self._factor) !=  self.pixel_n_dim:
+            raise ValueError(f"Length of factor must equal number of dimensions {self.pixel_n_dim}.")
+        if np.isscalar(offset):
+            offset = [offset] * self.pixel_n_dim
+        self._offset = np.array(offset)
+        if len(self._offset) != self.pixel_n_dim:
+            raise ValueError(f"Length of offset must equal number of dimensions {self.pixel_n_dim}.")
+
+    def _top_to_underlying_pixels(self, top_pixels):
+        # Convert user-facing pixel indices to the pixel grid of underlying WCS.
+        factor = self._pad_dims(self._factor, top_pixels.ndim)
+        offset = self._pad_dims(self._offset, top_pixels.ndim)
+        return top_pixels * factor + offset
+
+    def _underlying_to_top_pixels(self, underlying_pixels):
+        # Convert pixel indices of underlying pixel grid to user-facing grid.
+        factor = self._pad_dims(self._factor, underlying_pixels.ndim)
+        offset = self._pad_dims(self._offset, underlying_pixels.ndim)
+        return (underlying_pixels - offset) / factor
+
+    def _pad_dims(self, arr, ndim):
+        # Pad array with trailing degenerate dimensions.
+        # This make scaling with pixel arrays easier.
+        shape = np.ones(ndim, dtype=int)
+        shape[0] = len(arr)
+        return arr.reshape(tuple(shape))
 
     def pixel_to_world_values(self, *pixel_arrays):
-        pixel_arrays = [np.asarray(pixel_arrays[i]) * self._factor[i]
-                        for i in range(self.pixel_n_dim)]
-        return self._wcs.pixel_to_world_values(*pixel_arrays)
+        underlying_pixel_arrays = self._top_to_underlying_pixels(np.asarray(pixel_arrays))
+        return self._wcs.pixel_to_world_values(*underlying_pixel_arrays)
 
     def world_to_pixel_values(self, *world_arrays):
-        pixel_arrays = self._wcs.world_to_pixel_values(*world_arrays)
-        pixel_arrays = [np.asarray(pixel_arrays[i]) / self._factor[i]
-                        for i in range(self.pixel_n_dim)]
-        return pixel_arrays
+        underlying_pixel_arrays = self._wcs.world_to_pixel_values(*world_arrays)
+        top_pixel_arrays = self._underlying_to_top_pixels(np.asarray(underlying_pixel_arrays))
+        return tuple(array for array in top_pixel_arrays)
 
     @property
     def pixel_shape(self):
-        return tuple(self._wcs.pixel_shape[i] / self._factor[i]
-                     for i in range(self.pixel_n_dim))
+        # Return pixel shape of resampled grid.
+        # Where shape is an integer, return an int type as its required for some uses.
+        if self._wcs.pixel_shape is None:
+            return self._wcs.pixel_shape
+        underlying_shape = np.asarray(self._wcs.pixel_shape)
+        int_elements = np.isclose(np.mod(underlying_shape, self._factor), 0,
+                                  atol=np.finfo(float).resolution)
```

Comment:
:exploding_head:

Reply:
Actually
```python
        int_elements = np.isclose(underlying_shape / self._factor, np.rint(underlying_shape / self._factor),
                                  atol=np.finfo(float).resolution)
```
might be the most robust way to check this.